### PR TITLE
allow for pfdns to fork

### PR DESCRIPTION
**Pull request by julsemaan**
# Description

Allow for pfdns to fork to process multiple requests in parallel 
# Impacts

pfdns
# Delete branch after merge

YES
# NEWS file entries
## New Features
- The pfdns service now forks to process multiple requests in parallel

### CODE REVIEW

File: sbin/pfdns [hunk 1]
```diff
@@ -143,6 +143,33 @@ foreach my $network ( keys %ConfigNetworks ) {
...
+    my $pid = fork();
+    if($pid) {
+        $CHILDREN{$pid} = $id;
+        $SIG{CHLD} = "IGNORE";
+    } elsif ($pid == 0) {
+        $SIG{CHLD} = "DEFAULT";
+        $IS_CHILD = 1;
+        _run_child($id);
+    } else {
+    }
```
Comment (cgx) [inline sbin/pfdns]: Useless; remove.

**Comment by extrafu** [comment]
What is the performance difference between the single process approach and multiple processes.

Please do some tests and post results.

**Comment by louismunro** [comment]
This has been on hold for a while because bugs.
We should make sure they are fixed if it is to be merged at any point.

**Comment by julsemaan** [comment]
@louismunro I think you are confusing pfdns and pfconfig

**Comment by louismunro** [comment]
https://www.google.ca/url?sa=i&rct=j&q=&esrc=s&source=images&cd=&ved=0ahUKEwjNo76tgsrJAhUGyRQKHX78BUIQjRwIBw&url=https%3A%2F%2Farchive.4plebs.org%2Fs4s%2Fthread%2F2340408%2F&psig=AFQjCNFKglY5vBu-cs8Rf1uhwBsGal8qvA&ust=1449586951781220

**Comment by julsemaan** [comment]
Here are the test results from last week : 

Without forking : 

```
Statistics:

  Queries sent:         60000
  Queries completed:    53459
  Queries lost:         6541
  Run time (s):         100.000000
  Maximum throughput:   740.000000 qps
  Lost at that point:   18.68%
```

With forking : 

```
Statistics:

  Queries sent:         60000
  Queries completed:    57806
  Queries lost:         2194
  Run time (s):         100.000000
  Maximum throughput:   1646.000000 qps
  Lost at that point:   1.79%
```

Approximately 2.5 times more requests per second, lost reduced from ~19% to ~2%

This is with 4 pfdns processes answering requests.

Results could be better I think as the test environment had the client and server running on the same ESX.

File: sbin/pfdns [hunk 2]
```diff
@@ -343,6 +374,10 @@ sub isolzone {
 }
 
 sub normal_sighandler {
+    foreach my $pid (keys %CHILDREN){
+        kill(SIGKILL, $pid);
+    }
+    deletepid();
```
Comment (fdurand) [inline sbin/pfdns:380]: Hum you probably fix a bug with that.

**Comment by fdurand** [comment]
Looks good to me, merge

File: sbin/pfdns (hunk 2 above)
Comment (julsemaan) [inline sbin/pfdns:380]: Unintentional bug fixing :)

Life just gives out bonuses like these

**Pull request merged**